File: start-server.py
#!/usr/bin/env python3
"""
Simple HTTP server for local development
Usage: python start-server.py [port]
Default port: 8000
"""

import http.server
import socketserver
import sys
import webbrowser
import os
from pathlib import Path

def start_server(port=8000):
    """Start the HTTP server"""
    
    # Change to the script's directory (project root)
    script_dir = Path(__file__).parent
    os.chdir(script_dir)
    
    class CustomHandler(http.server.SimpleHTTPRequestHandler):
        def do_GET(self):
            if self.path == '/':
                self.path = '/src/index.html'
            return super().do_GET()
    
    Handler = CustomHandler
    
    try:
        with socketserver.TCPServer(("", port), Handler) as httpd:
            print(f"🚀 Server started at http://localhost:{port}")
            print(f"📁 Serving directory: {script_dir}")
            print("Press Ctrl+C to stop the server")
            
            # Open browser automatically
            webbrowser.open(f"http://localhost:{port}")
            
            httpd.serve_forever()
            
    except KeyboardInterrupt:
        print("\n👋 Server stopped")
    except OSError as e:
        if e.errno == 48:  # Address already in use
            print(f"❌ Port {port} is already in use. Try a different port:")
            print(f"   python start-server.py {port + 1}")
        else:
            print(f"❌ Error starting server: {e}")
        sys.exit(1)

if __name__ == "__main__":
    port = 8000
    
    if len(sys.argv) > 1:
        try:
            port = int(sys.argv[1])
        except ValueError:
            print("❌ Invalid port number. Using default port 8000")
            port = 8000
    
    if port < 1 or port > 65535:
        print("❌ Port must be between 1 and 65535")
        sys.exit(1)
    
    start_server(port)

File: start-server.js
#!/usr/bin/env node
/**
 * Simple HTTP server for local development
 * Usage: node start-server.js [port]
 * Default port: 8000
 */

const http = require('http');
const fs = require('fs');
const path = require('path');

const PORT = parseInt(process.argv[2], 10) || 8000;
const ROOT = __dirname;

const MIME_TYPES = {
  '.html': 'text/html; charset=utf-8',
  '.css': 'text/css; charset=utf-8',
  '.js': 'application/javascript; charset=utf-8',
  '.json': 'application/json; charset=utf-8',
  '.png': 'image/png',
  '.ico': 'image/x-icon',
  '.svg': 'image/svg+xml',
  '.md': 'text/markdown; charset=utf-8',
  '.txt': 'text/plain; charset=utf-8',
};

const server = http.createServer((req, res) => {
  let filePath = path.join(ROOT, req.url === '/' ? 'index.html' : req.url);
  const ext = path.extname(filePath);

  fs.readFile(filePath, (err, data) => {
    if (err) {
      res.writeHead(404, { 'Content-Type': 'text/plain' });
      res.end('404 Not Found');
      return;
    }

    res.writeHead(200, {
      'Content-Type': MIME_TYPES[ext] || 'application/octet-stream',
      'Cache-Control': 'no-cache',
    });
    res.end(data);
  });
});

server.listen(PORT, () => {
  const url = `http://localhost:${PORT}`;
  console.log(`\n  🚀  Server running at ${url}`);
  console.log(`  📁  Serving: ${ROOT}`);
  console.log(`  ⌨️   Press Ctrl+C to stop\n`);

  // Try to open browser (cross-platform)
  const { exec } = require('child_process');
  const platform = process.platform;
  const cmd =
    platform === 'darwin' ? `open ${url}` :
    platform === 'win32' ? `start ${url}` :
    `xdg-open ${url}`;
  exec(cmd, () => {});
});

server.on('error', (err) => {
  if (err.code === 'EADDRINUSE') {
    console.error(`\n  ❌  Port ${PORT} is already in use. Try a different port:`);
    console.error(`     node start-server.js ${PORT + 1}\n`);
  } else {
    console.error(`\n  ❌  Error: ${err.message}\n`);
  }
  process.exit(1);
});
